### PR TITLE
[WIP] Fix #2022 :  Add feature to allow host to toggle PhaseSplit Visibility on Leaderboard tab

### LINKED ISSUE #2022: Add feature to toggle leaderboard visibility by challenge host
## Current Behaviour

The challenge host can view the challenge leaderboard but can't make it public/private after the challenge is completed.

## Expected Behaviour

A challenge host will be able to make the leaderboard public/private using the UI.

### Backend
- [ ] Add a check in `get_or_update_challenge_phase_split` API for PATCH requests that the requesting user should a challenge host or a member of challenge host team.
- [ ] Add two tests one when the requesting user is a member of host team and one if the user isn't the member of host team.

### Frontend
- [ ] Add a toggle button in the leaderboard tab of the challenge host which will appear when the leaderboard is selected.
- [ ] Add a function in the challenge controller to which will be bind to the above toggle button to change the visibility.

Please feel free to ask if you have some doubts!

### PR DESCRIPTION
**Pull request by nileshprasad137**
Fixes #2022 

### Backend
  - [x] Add a check in get_or_update_challenge_phase_split API for PATCH requests that the requesting user should a challenge host or a member of challenge host team.
  - [x] Add two tests one when the requesting user is a member of host team and one if the user isn't the member of host team.
### Frontend
  - [x] Add a toggle button in the leaderboard tab of the challenge host which will appear when the leaderboard is selected.
  - [x] Add a function in the challenge controller to which will be bind to the above toggle button to change the visibility.

### CODE REVIEW

**Comment by nileshprasad137** [comment]
Can anyone help me understand the issue with Travis build here. After my first commit, there were some PEP8 linting issues but all the tests passed. Refer [this Travis build](https://travis-ci.org/Cloud-CV/EvalAI/builds/487130406?utm_source=github_status&utm_medium=notification). So, I resolved it via second commit. This time there were no PEP8 issues but 1 testcase failed in `GetRemainingSubmissionTest.test_get_remaining_submission_when_submission_made_three_days_back `. 

Why did it happen? I've not made any changes inside `Jobs` unit tests.

**Comment by vkartik97** [comment]
Hi @nileshprasad137 
I am looking into the Travis Build Issue.
All the Travis PR Submission after #2057 is facing same issue, so no need to worry!
Please let me know if you have questions. :)

File: apps/challenges/views.py [hunk 1]
```diff
@@ -1255,16 +1258,30 @@ def get_or_update_challenge_phase_split(request, challenge_phase_split_pk):
     """
     challenge_phase_split = get_challenge_phase_split_model(
         challenge_phase_split_pk)
+    challenge_phase_id = ChallengePhaseSplit.objects.filter(pk=challenge_phase_split_pk).values_list(
+        'challenge_phase_id', flat=True)[0]
+    challenge_id = ChallengePhase.objects.filter(pk=challenge_phase_id).values_list('challenge_id', flat=True)[0]
+
```
Comment (RishabhJain2018) [inline apps/challenges/views.py]: @nileshprasad137 There is no need to query the database again. You can simply get the challenge id from challenge phase using `challenge_id = challenge_phase_split.challenge_phase.challenge.id`

File: apps/challenges/views.py [hunk 2]
```diff
@@ -1255,16 +1258,30 @@ def get_or_update_challenge_phase_split(request, challenge_phase_split_pk):
...
+    # Check if user is a challenge host
+    is_user_host = is_user_a_host_of_challenge(request.user, challenge_id)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py]: `is_user_host` -> `is_host_user`

File: apps/challenges/views.py [hunk 3]
```diff
@@ -1255,16 +1258,30 @@ def get_or_update_challenge_phase_split(request, challenge_phase_split_pk):
...
+
+    # Check if user is a member of challenge host team
+    host_team_ids = get_challenge_host_teams_for_user(request.user)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py]: `host_team_ids` -> `host_team_list`

File: apps/challenges/views.py [hunk 4]
```diff
@@ -1255,16 +1258,30 @@ def get_or_update_challenge_phase_split(request, challenge_phase_split_pk):
...
     if request.method == 'PATCH':
-        serializer = ZipChallengePhaseSplitSerializer(challenge_phase_split,
-                                                      data=request.data,
-                                                      partial=True)
-        if serializer.is_valid():
-            serializer.save()
-            response_data = serializer.data
-            return Response(response_data, status=status.HTTP_200_OK)
-        return Response(serializer.erros, status=status.HTTP_400_BAD_REQUEST)
+        if is_user_host or is_user_member_of_host_team:
```
Comment (RishabhJain2018) [inline apps/challenges/views.py]: `is_user_host` -> `is_host_user`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 5]
```diff
@@ -1643,6 +1644,86 @@
...
+                    onError: function(response) {
+                        $mdDialog.hide();
+                        vm.page.description = vm.tempDesc;
+                        var error = response.data;
+                        $rootScope.notify("error", error);
+                    }
+                };
+
+                utilities.sendRequest(parameters);
+            }, function() {
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js]: Please remove this function.

File: frontend/src/js/controllers/challengeCtrl.js [hunk 6]
```diff
@@ -1643,6 +1644,86 @@
...
+            // Nope
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js]: Please remove this comment.

File: apps/challenges/views.py (hunk 1 above)
Comment (nileshprasad137) [reply apps/challenges/views.py]: Oh yes! I missed it. I'll fix them.

**Comment by nileshprasad137** [comment]
Did the requested changes.

**Comment by RishabhJain2018** [comment]
Hi @nileshprasad137 Can you please resolve the conflicts?
@fristonio Please review this PR.

**Pull request closed without merging**